homeassistant/components/monzo/services.py

```diff
@@ -0,0 +1,153 @@
+
+
+def _get_api(
+    call: ServiceCall, device_registry: DeviceRegistry, hass: HomeAssistant
+) -> AuthenticatedMonzoAPI:
+    """Get the API from the config entry associated with the chosen pot."""
+    first_pot = device_registry.async_get(call.data[TRANSFER_POTS][0])
+    if first_pot is None:
+        raise ServiceValidationError(
+            translation_domain=DOMAIN,
+            translation_key="invalid_device",
+            translation_placeholders={"device_id": call.data[TRANSFER_POTS][0]},
+        )
+    entry_id = next(iter(first_pot.config_entries))
+    coordinator: MonzoCoordinator = hass.data[DOMAIN][entry_id]
+    return coordinator.api
+
+
+async def _get_account_id(
+    call: ServiceCall, api: AuthenticatedMonzoAPI, device_registry: DeviceRegistry
+) -> str:
+    """Get the Monzo account ID from the device, defaulting to current account."""
+    return (
+        await _get_current_account_id(api)
+        if TRANSFER_ACCOUNT not in call.data
+        else await _get_account_id_from_device(
+            device_registry, call.data[TRANSFER_ACCOUNT], VALID_TRANSFER_ACCOUNTS
+        )
+    )
+
+
+async def _get_current_account_id(api: AuthenticatedMonzoAPI) -> str:
+    """Get the current account ID from the Monzo API."""
+    curent_account_id: str = next(
+        acc["id"]
+        for acc in (await api.user_account.accounts())
+        if acc["type"] == "uk_retail"
+    )
+
+    return curent_account_id
+
+
+async def _get_account_id_from_device(
+    device_registry: DeviceRegistry, device_id: str, valid_account_types: set[str]
+) -> str:
+    """Get the Monzo account ID for a given device."""
+    device = await _get_device(device_registry, device_id)
+    if device.model not in valid_account_types:
+        raise ServiceValidationError(
+            translation_domain=DOMAIN,
+            translation_key="invalid_device_model",
+            translation_placeholders={
+                "device_name": device.name or device_id,
+                "valid_types": str(valid_account_types),
+            },
+        )
+    _, account_id = next(iter(device.identifiers))
+    return account_id
+
+
+async def _get_pot_ids(call: ServiceCall, device_registry: DeviceRegistry) -> list[str]:
+    """Get the Monzo account IDs for the selected pots in the ServiceCall."""
+    return await asyncio.gather(
+        *[
+            _get_account_id_from_device(device_registry, acc, VALID_POT_ACCOUNTS)
+            for acc in call.data.get(TRANSFER_POTS, "")
+        ]
+    )
+
+
+async def _get_device(device_registry: DeviceRegistry, device_id: str) -> DeviceEntry:
+    """Get a device from the DeviceRegistry or raise a ServiceValidationError."""
+    device = device_registry.async_get(device_id)
+    if not device:
+        raise ServiceValidationError(
+            translation_domain=DOMAIN,
+            translation_key="invalid_device",
+            translation_placeholders={"device_id": device_id},
+        )
+    return device
```

Comment:
I dislike that they almost all need the device_registry and the service call, it makes them feel like methods you can use for this service call.